pkg/oauthserver/http2/misdirected_request.go

```diff
@@ -0,0 +1,226 @@
+// and resolve to the same IPs through DNS, an existing HTTP2 server connection is considered reusable for the other
+// servers' packets. However, because the HTTP2 connection at the proxy is coupled to the HTTP2 server through the
+// initial SNI header from a TLS handshake, and the packets coming through are opaque and cannot be disambiguated by
+// the proxy, the packets cannot be identified by the proxy as misdirected and are all forwarded to the HTTP2 server.
+//
+// Solutions could be:
+//
+// 1. Discontinue use of HTTP2 for these services
+// 2. Implement mutual TLS in the ingress controller to enable terminating TLS at the proxy for the auth server
+// 3. Implement HTTP 421 misdirected request support at the auth server to hint clients to stop reusing the connection for the request authority
+//
+// Our current recommendation in the short term is (3), and longer term we would like to implement mTLS (2) for ease of use.
+//
+// References:
+//
+// * https://httpwg.org/specs/rfc7540.html#reuse
+// * https://httpwg.org/specs/rfc7540.html#MisdirectedRequest
+// * https://daniel.haxx.se/blog/2016/08/18/http2-connection-coalescing
+
+const (
+	// Explicitly limit to a single concurrent open stream per client in
+	// an attempt to reduce the chance that we get misdirected requests.
+	MaxStreamsPerConnection = 1
+
+	http2ProtoMajor = 2
+
+	responseMisdirectedRequest = `
+<html xmlns="http://www.w3.org/1999/xhtml">
+  <head>
+    <meta http-equiv="refresh" content="0" />
+  </head>
+  <body>
+    <p>misdirected request</p>
+  </body>
+</html>`
+)
+
+func WithMisdirectedRequest(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// happy path
+		if !isMisdirectedRequest(r) {
+			handler.ServeHTTP(w, r)
+
+			// stop processing the request
+			return
+		}
+
+		// log only the safe metadata for this misdirected request
+		klog.Infof("misdirected request detected from %s to %s %s %d instead of %s",
+			r.RemoteAddr, r.Method, r.Host, r.ContentLength, r.TLS.ServerName)
+
+		// send the client a graceful close connection via GOAWAY
+		w.Header().Set("Connection", "close")
+
+		// set 421 code so that well behaved clients retry the request
+		w.WriteHeader(http.StatusMisdirectedRequest)
+
+		// try to force a browser refresh for misbehaving clients
+		_, _ = w.Write([]byte(responseMisdirectedRequest))
+	})
+}
+
+func isMisdirectedRequest(r *http.Request) bool {
+	// ignore non-HTTP2 requests
+	if r.ProtoMajor != http2ProtoMajor {
+		return false
+	}
+
+	// check if have a value for the :authority pseudo header field
+	// see golang.org/x/net/http2/server.go serverConn.newWriterAndRequest / newWriterAndRequestNoBody
+	host := r.Host
+	if len(host) == 0 {
+		return false
+	}
+
+	// ignore non-TLS requests
+	tls := r.TLS
+	if tls == nil {
+		return false
+	}
+
+	// check if we know the server name requested by the client
+	serverName := tls.ServerName
+	if len(serverName) == 0 {
```

Comment:
In what situation would this be 0? A non-SNI request?

Reply:
`ServerName                  string                // server name requested by client, if any (server side only)`

I think it is optional in general?  For example, I believe when the operator hits `/healthz` this value is empty.